### PR TITLE
Add equal input check while merging payloads

### PR DESCRIPTION
**Pull request by GuanLuo**
Before this change, the payload merging may be performed without checking the shape consistency. This change extend the same check in batcher to the merging function for completeness.

@tanmayv25 I think this is needed to fix the root cause. That being said, I will still revisit the GetDynamicBatch() logic for better batching: we should prefer growing existing payload over creating new payload because we have limit on payloads enqueued (2 payload / instance). If 4 requests arrive, forming 2 2-request payloads has different result from 4 1-request payloads even if payload merging is performed.

### CODE REVIEW

File: src/payload.cc [hunk 1]
```diff
@@ -25,6 +25,7 @@
 // OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 #include "payload.h"
+#include "triton/common/logging.h"
```
Comment (dyastremsky) [inline src/payload.cc:28]: Is this import necessary? It doesn't look you added any logging calls.

Comment (GuanLuo) [reply src/payload.cc:28]: This is from debugging, will remove

**Review by rmccorm4 — COMMENTED** [review]
Logic and encapsulation make sense to me.

To summarize, is this basically moving the shape check/tracking into Payload class (since it all goes through Payload anyways), and also performing this check in PayloadMerge in addition to existing checks in scheduler? 

I'll leave approval to Tanmay

**Review by tanmayv25 — APPROVED** [review]
The performance impact will be only for the cases where the timeout was encountered. It looks okay for resolving the issue.

**Review by rmccorm4 — APPROVED** [review]

**Review by Tabrizian — COMMENTED** [review]
Do we need a test for this to make sure we have coverage in L0_batcher? Just to make sure I'm understanding the issue correctly, before this change we could be creating batches with inconsistent shapes?

**Comment by tanmayv25** [comment]
Yes. Our current tests don't capture the timeout cases where this merge can happen.

**Pull request merged**